runtime/Cpp/runtime/src/atn/PredictionContext.h

```diff
@@ -217,36 +224,62 @@ namespace atn {
     std::vector<std::string> toStrings(Recognizer *recognizer, const Ref<PredictionContext> &stop, int currentState);
   };
 
+  struct PredictionContextPair {
+    PredictionContextPair(const Ref<PredictionContext>& lhs_, const Ref<PredictionContext>& rhs_);
+    PredictionContextPair(PredictionContextPair&& other);
+
+    Ref<PredictionContext> lhs;
+    Ref<PredictionContext> rhs;
+
+    bool operator == (const PredictionContextPair& o) const;
+  };
+
   struct PredictionContextHasher {
     size_t operator () (const Ref<PredictionContext> &k) const {
       return k->hashCode();
     }
   };
 
   struct PredictionContextComparer {
-    bool operator () (const Ref<PredictionContext> &lhs, const Ref<PredictionContext> &rhs) const
-    {
+    bool operator () (const Ref<PredictionContext> &lhs, const Ref<PredictionContext> &rhs) const {
       if (lhs == rhs) // Object identity.
         return true;
       return (lhs->hashCode() == rhs->hashCode()) && (*lhs == *rhs);
     }
   };
+}
+}
+
+namespace std {
+  template <>
+  struct hash<antlr4::atn::PredictionContextPair> {
+    size_t operator()(const antlr4::atn::PredictionContextPair& pair) const {
+      return pair.lhs->cachedHashCode ^ pair.rhs->cachedHashCode;
+    }
+  };
+}
+
+namespace antlr4 {
+namespace atn {
 
   class PredictionContextMergeCache {
   public:
-    Ref<PredictionContext> put(Ref<PredictionContext> const& key1, Ref<PredictionContext> const& key2,
-                               Ref<PredictionContext> const& value);
-    Ref<PredictionContext> get(Ref<PredictionContext> const& key1, Ref<PredictionContext> const& key2);
+    PredictionContextMergeCache();
+
+    void put(Ref<PredictionContext> const& key1,
+             Ref<PredictionContext> const& key2,
+             Ref<PredictionContext> const& value);
+    // This is slightly different to the Java runtime. get is a symmetric lookup.
+    const Ref<PredictionContext>& get(Ref<PredictionContext> const& key1,
+                                      Ref<PredictionContext> const& key2) const;
 
     void clear();
     std::string toString() const;
     size_t count() const;
 
   private:
-    std::unordered_map<Ref<PredictionContext>,
-      std::unordered_map<Ref<PredictionContext>, Ref<PredictionContext>, PredictionContextHasher, PredictionContextComparer>,
-      PredictionContextHasher, PredictionContextComparer> _data;
-
+    Ref<PredictionContext> _missing;
+    std::unordered_map<PredictionContextPair, Ref<PredictionContext>> _data;
```

Comment:
It seems that I have to add copy constructor to PredictionContextPair `PredictionContextPair::PredictionContextPair(const PredictionContextPair& o): lhs(o.lhs), rhs(o.rhs) { }` to make `std::unordered_map<PredictionContextPair, Ref<PredictionContext>>` works for MSVC2017?